src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkWasmModule.java

```diff
@@ -0,0 +1,278 @@
+    }
+
+    int inputLen = Math.toIntExact(input.length);
+    int inputPtr = alloc(allocateFn, allocFn, inputLen, 1);
+    try (SilentCloseable c = prof.profile(WASM_EXEC, "copy input")) {
+      memory.write(inputPtr, input);
+    }
+
+    // struct { output_ptr_ptr: **u8, output_len_ptr: *u32 }
+    int paramsPtr = alloc(allocateFn, allocFn, 8, 4);
+    int outputPtrPtr = paramsPtr;
+    int outputLenPtr = paramsPtr + 4;
+    memory.writeI32(outputPtrPtr, 0);
+    memory.writeI32(outputLenPtr, 0);
+
+    long[] execResult;
+    try (SilentCloseable c = prof.profile(WASM_EXEC, "execute")) {
+      execResult = execFn.apply(inputPtr, inputLen, outputPtrPtr, outputLenPtr);
+    }
+
+    // FIXME: Not 100% sure this check is necessary, but the ambiguity between
+    // signed/unsigned in Java vs WebAssembly makes me nervous.
+    //
+    // Might be unnecessary if the function signature is verified before execution?
+    long returnCode = execResult[0];
+    if (returnCode < 0 || returnCode > 0xFFFFFFFFL) {
+      returnCode = 0xFFFFFFFFL;
+    }
+    int outputPtr = memory.readInt(outputPtrPtr);
+    int outputLen = memory.readInt(outputLenPtr);
+
+    String output = "";
+    if (outputLen > 0) {
+      try (SilentCloseable c = prof.profile(WASM_EXEC, "copy output")) {
+        byte[] outputBytes = memory.readBytes(outputPtr, outputLen);
+        output = new String(outputBytes, ISO_8859_1);
+      }
+    }
+    return StarlarkWasmExecutionResult.newOk(returnCode, output);
+  }
+
+  private static void validateModule(WasmModule wasmModule, String allocateFn) throws EvalException {
+    var exports = wasmModule.exportSection();
+    int exportCount = exports.exportCount();
+    for (int ii = 0; ii < exportCount; ii++) {
+      var export = exports.getExport(ii);
+      if (export.name().equals(allocateFn)) {
+        // FIXME: Validate exported type is a function and has the expected signature?
+        return;
+      }
+    }
+    throw Starlark.errorf("WebAssembly module doesn't contain an export named \"%s\"", allocateFn);
+  }
+
+  MemoryLimits getMemLimits(long memLimitBytes) {
+    int initialPages = 1;
+    int memLimitPages = getMemLimitPages(memLimitBytes);
+
+    if (wasmModule.memorySection().isPresent()) {
+      var memories = wasmModule.memorySection().get();
+      int memoryCount = memories.memoryCount();
+      for (int ii = 0; ii < memoryCount; ii++) {
+        MemoryLimits limits = memories.getMemory(ii).limits();
+        if (limits.initialPages() > initialPages) {
+          initialPages = limits.initialPages();
+        }
+      }
+    }
+    if (initialPages > memLimitPages) {
+      // FIXME: Should probably throw an exception. The execution will likely fail anyway, and
+      // throwing an exception from this point would provide more relevant details.
+      initialPages = memLimitPages;
+    }
+    return new MemoryLimits(initialPages, memLimitPages);
+  }
+
+  static int getMemLimitPages(long memLimitBytes) {
```

Comment:
this function can be replaced with `Math.min(MAX_PAGES, Math.ceilDiv(memLimitBytes, PAGE_SIZE))`. IMO it's also more readable that way, as it clearly describes the intent.

Reply:
Done